### PR TITLE
WIP Validate lineage for all JexlNode rebuilding visitors

### PR DESCRIPTION
**Pull request by lbschanno**
WIP for #880 for verifying/fixing JexlNode rebuilding visitors to ensure that they preserve proper parentage. 

I have fixed/verified parentage for the following visitors:

* `AllTermsIndexedVisitor`
* `BooleanOptimizationVisitor`
* `FacetCheck`
* `DateIndexCleanupVisitor`
* `ExpandMultinormalizedTerms`
* `TreeFlatteningRebuildingVisitor`
* `FixNegativeNumbersVisitor`
* `GeowavePruningVisitor`
* `QueryModelVisitor`
* `RangeCoalescingVisitor`
* `RegexFunctionVisitor`

I am currently working on fixing parentage for the following visitors:

* `UniqueExpressionTermsVisitor`
* `ExpandCompositeTerms`

There are a number of rebuilding visitors that do not have any unit tests associated with them that I could find, nor was there enough documentation for me to be able to extrapolate test cases for them. I would appreciate any help with examples or desired test cases so that I can write tests for each of the following and verify parentage.

* `BoundedRangeDetectionVisitor`
* `FunctionIndexQueryExpansionVisitor`
* `FunctionNormalizationRebuildingVisitor`
* `IsNotNullIntentVisitor`
* `ParallelIndexExpansion`
* `PruneLessSelectiveFieldsVisitor`
* `PushdownLargeFieldedListsVisitor`
* `PushdownMissingIndexRangeNodesVisitor`
* `PushFunctionsIntoExceededValueRanges`
* `RangeConjunctionRebuildingVisitor`
* `RangeExpansionThresholdRebuildingVisitor`

### CODE REVIEW

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java [hunk 1]
```diff
@@ -63,143 +67,91 @@ public AllTermsIndexedVisitor(ShardQueryConfiguration config, MetadataHelper hel
...
+        // @formatter:off
+        JexlNode[] children = Arrays.stream(JexlNodes.children(node))
+                        .map(n -> (JexlNode) n.jjtAccept(this, data)) // Visit the node
+                        .filter(Objects::nonNull) // Filter out null nodes
+                        .filter(n -> (!JexlNodes.isAnd(n) && !JexlNodes.isOr(n)) || JexlNodes.isNotChildless(n)) // Filter out empty junction nodes
+                        .toArray(JexlNode[]::new);
+        // @formatter:on
+        JexlNode copy = JexlNodes.newInstanceOfType(node);
+        copy.image = node.image;
+        System.out.println("Copy: " + copy);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java:116]: Please remove System.out.println.  Use the logger if anything.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java [hunk 2]
```diff
@@ -63,143 +67,91 @@ public AllTermsIndexedVisitor(ShardQueryConfiguration config, MetadataHelper hel
...
                 return child;
             default:
-                return newNode;
+                return copy;
         }
     }
     
+    private JexlNode getCopyWithVisitedChildren(JexlNode node, Object data) {
+        // @formatter:off
+        JexlNode[] children = Arrays.stream(JexlNodes.children(node))
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java:108]: nice

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BooleanOptimizationRebuildingVisitor.java [hunk 3]
```diff
@@ -54,27 +50,22 @@ protected JexlNode optimizeTree(JexlNode currentNode, JexlNode newNode, Object d
...
-            
             JexlNode toAttach = nodes.second();
             toAttach = TreeFlatteningRebuildingVisitor.flatten(toAttach);
+            ASTOrNode newNode = new ASTOrNode(ParserTreeConstants.JJTORNODE);
+            newNode.image = currentNode.image;
             
             for (int i = 0; i < toAttach.jjtGetNumChildren(); i++) {
                 JexlNode node = copy(prunedNode);
                 JexlNode attach = (JexlNode) toAttach.jjtGetChild(i).jjtAccept(this, data);
+                attach.jjtSetParent(node);
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BooleanOptimizationRebuildingVisitor.java:61]: I see lots of refactoring here but this is the only line that actually fixes a parenting problem for this class.  Refactoring for the sake of refactoring may do more harm than good if anybody else is working on these visitors for any other reason.  It also makes it very difficult for anybody to actually accept the entire package and merge.  I think you may need to pull back a little from the refactoring efforts.  Better yet a separate pull request per visitor that is only refactoring would be considerably more palatable.

**Review by ivakegg — CHANGES_REQUESTED** [review]

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/QueryModelVisitor.java [hunk 4]
```diff
@@ -1,16 +1,14 @@
...
-import com.google.common.collect.FluentIterable;
 import com.google.common.collect.Lists;
 import com.google.common.collect.Multimap;
 import com.google.common.collect.Sets;
-import datawave.query.model.QueryModel;
 import datawave.query.exceptions.DatawaveFatalQueryException;
 import datawave.query.jexl.JexlASTHelper;
 import datawave.query.jexl.JexlNodeFactory;
 import datawave.query.jexl.JexlNodeFactory.ContainerType;
+import datawave.query.model.QueryModel;
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/QueryModelVisitor.java:11]: OK, I am stumped in this visitor as to where any lineage problems were fixed.  Is this one purely refactoring?

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RangeCoalescingVisitor.java [hunk 5]
```diff
@@ -90,12 +50,12 @@ public Object visit(ASTAndNode node, Object data) {
             JexlNodes.ensureCapacity(andNode, node.jjtGetNumChildren());
             for (int i = 0; i < node.jjtGetNumChildren(); i++) {
                 JexlNode newChild = (JexlNode) node.jjtGetChild(i).jjtAccept(this, data);
-                
                 andNode.jjtAddChild(newChild, i);
                 newChild.jjtSetParent(andNode);
             }
         }
-        
+
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RangeCoalescingVisitor.java:57]: This visitor looks good.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RegexFunctionVisitor.java [hunk 6]
```diff
@@ -48,55 +50,59 @@ public Object visit(ASTFunctionNode node, Object data) {
...
-                JexlNode regexNode = buildRegexNode((ASTIdentifier) node0, functionMetadata.name(), arguments.get(1).image);
-                if (regexNode != null) {
-                    returnNode = regexNode;
-                }
-            } else {
-                JexlNode newParent = null;
-                if (functionMetadata.name().equals("excludeRegex")) {
-                    newParent = new ASTAndNode(ParserTreeConstants.JJTANDNODE);
+    
+        switch (functionMetadata.name()) {
```
Comment (ivakegg) [inline warehouse/query-core/src/main/java/datawave/query/jexl/visitors/RegexFunctionVisitor.java:54]: much cleaner, thanks

**Comment by ivakegg** [comment]
First I must explain what a (query property) "marker" node is.  A portion of a query (jexl subtree) can be marked as have some attribute by attaching an assignment node to that subtree.  So if I want to mark some node A with a marker N, then I replace A with ((N = true) && (A)).  These markers can be thought of as query execution hints that the planner puts in place to allow the execution to do certain things.

Second I must explain what "push down" means.  To "push down" some portion of a query (jexl subtree) is to use the field index instead of the global index, or pushing down farther to the query evaluation phase.  Query execution basically started with global index lookup, then field index lookup (if required), then query evaluation meaning bouncing the boolean logic against all of the values for a document.

Here is some description of these visitors:

- BoundedRangeDetectionVisitor
This is supposed to detect whether there are bounded ranges, however it appears to only be looking for bounds.  I am actually changing this visitor here: https://github.com/NationalSecurityAgency/datawave/pull/926
- FunctionIndexQueryExpansionVisitor
This will expand all of the functions within the query to include something that we can run against the index.  Basically every function node in the tree will be replaced with (function && (<some query)).  The index query created will be dependent on the query descriptor for the query in question.
- FunctionNormalizationRebuildingVisitor
This visitor will normalize the values within the arguments of a function node.  The reason this is separated out is that the function descriptor needs to be used to determine which arguments are actually values and for which field do they apply.  The field name is needed to lookup the data type and subsequently the normalization required.  Normalization of values is required for index lookup.
- IsNotNullIntentVisitor
if an ER (regular expression) node is of the form FIELD =~ '*', then replace with: FIELD != null
- ParallelIndexExpansion
This is used for expanding regex nodes into the set of discrete values.  It is also used for unfielded term expansion (FixUnfieldedTermsVisitor) by extension.  If a regex (or range) cannot be expanded into descrete values, then it is wrapped with an ExceededValueThresholdMarkerJexlNode marker.
- PruneLessSelectiveFieldsVisitor
Not used for now.  You can ignore this one for now until we start populating the metadata with more extensive metrics at which time this visitor may be changed extensively.
- PushdownLargeFieldedListsVisitor
This will take a very large list of values OR'ed together for one field into an ExceededOrThresholdMarkerJexlNode construct.  This essentially will allow an "Ivarator" (DatawaveFieldIndedListIteratorJexl) to be used instead of creating a separate source for each term (see QueryIterator).
- PushdownMissingIndexRangeNodesVisitor
This will push down terms as "evaluation only" (meaning the index is not used) when there is an entry in the metadata table that denotes that the field is not indexed for some time range.  This is an initial form of a solution for #825.
- PushFunctionsIntoExceededValueRanges
If we have an ExceededValueThresholdMarkerJexlNode on a range which had been produced as a result of the FunctionIndexQueryExpansionVisitor above, then the function is pushed into the marker node construct to allow the "Ivarator" to evaluation the function when scanning the field index.
- RangeConjunctionRebuildingVisitor
This will expand ranges into discrete values as the ParallelIndexExpansion visitor does for regex nodes.
- RangeExpansionThresholdRebuildingVisitor
This one is not used and can be ignored/dropped.

File: warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java (hunk 1 above)
Comment (lbschanno) [reply warehouse/query-core/src/main/java/datawave/query/jexl/visitors/AllTermsIndexedVisitor.java]: Fixed.

**Comment by lbschanno** [comment]
@ivakegg I have removed any refactoring not related to fixes. I will create separate PRs for that work as we discussed so that this PR is focused on fixes only.

**Comment by lbschanno** [comment]
Closing this PR to break it up into individual PRs for each visitor.

**Pull request closed without merging**